Improvements to Registry Entry Attributes

Leo40Git opened the pull request:
This PR adds the following improvements to the Registry Entry Attributes API:
- Added validation to all values, computed or otherwise
  - (done by trying to encode the value)
- `RegistryEntryAttribute` refactors!
  - Added `getNullable`, for hot code paths
  - Renamed `getValue` to `get`
    - Old name still exists, but is marked for removal
  - Added iteration methods!
    - `keySet` and `tagKeySet`, to get registry entries/tags that have an associated value
    - `entryIterator` and `tagEntryIterator`, to get pairs of registry entries/tags and their associated values
  - Added "value associated with entry" and "value associated with tag" events
    - Only fired for built-in values (`put`) - for data values, use `ResourceLoaderEvents.END_DATA_PACK_RELOAD`
- Merged data and assets holders
  - These were mutually exclusive anyway
- Added flags system to built-in holder
  - Right now, this is only used to exclude computed values (from `DefaultValueProvider`) from the `dump_builtin_attributes` command
- Marked `RegistryExtensions.register(Registry, Identifier, Object, Consumer<BuiltinAttachmentBuilder>)` and `BuiltinAttachmentBuilder` for removal
  - Nobody used these, and they've pretty much been replaced with the direct `RegistryEntryAttachment.put` method

Additionally, this PR also addresses two bugs:
- Fixed tag-based value associations not being cleared on datapack reload
- Fixed tag-based value associations not being cleared on synchronization from server

Leo40Git committed:
RegistryEntryAttachment: add "value associated with entry/tag" events

Leo40Git committed:
Merge data and asset holders

Leo40Git committed:
RegistryEntryAttachmentHolder: fix clear not clearing tag table

Leo40Git committed:
Add flags system to builtin holder

Currently only used to exclude computed values (from DefaultValueProvider)

Leo40Git committed:
Store values derived from tags

Leo40Git committed:
no grammar mistakes here, officer

Leo40Git committed:
forgot license on new file as usual

Leo40Git committed:
RegistryEntryAttachment.putValue: validate values

Leo40Git committed:
RegistryExtensions: don't bypass validation

OroArmor commented on library/data/registry_entry_attachments/src/main/java/org/quiltmc/qsl/registry/attachment/impl/BuiltinAttachmentBuilderImpl.java:
```diff
@@ -18,24 +18,20 @@
 import org.quiltmc.qsl.registry.attachment.api.RegistryExtensions;
 
 @ApiStatus.Internal
 public final class BuiltinAttachmentBuilderImpl<R> implements RegistryExtensions.BuiltinAttachmentBuilder<R> {
```
Maybe make this class deprecated or remove it? Doesnt do much anymore.

OroArmor commented on library/data/registry_entry_attachments/src/main/java/org/quiltmc/qsl/registry/attachment/impl/RegistryEntryAttachmentHolder.java:
```diff
@@ -134,7 +129,12 @@ public <T> void putValue(RegistryEntryAttachment<R, T> attachment, TagKey<R> tag
+
 	public void clear() {
 		this.valueTable.clear();
+		this.valueTagTable.clear();
```
oops

Leo40Git replied:
Hey, at least we noticed it before any end users :P

Leo40Git committed:
RegistryExtensions: remove BuiltinAttachmentBuilder API

Pretty much replaced with the direct RegistryEntryAttachment.put method and the other methods in RegistryExtensions

Leo40Git committed:
Javadoc and general cleanup

Leo40Git committed:
Fix computed flag persisting after value is replaced

Leo40Git committed:
fix comment

LambdAurora commented on library/data/registry_entry_attachments/src/main/java/org/quiltmc/qsl/registry/attachment/impl/RegistryEntryAttachmentImpl.java:
```diff
@@ -98,12 +107,27 @@ public Optional<V> getValue(R entry) {
+
+	@Override
+	public Event<ValueAdded<R, V>> valueAddedEvent() {
+		return valueAddedEvent;
```
```suggestion
		return this.valueAddedEvent;
```

LambdAurora commented on library/data/registry_entry_attachments/src/main/java/org/quiltmc/qsl/registry/attachment/impl/RegistryEntryAttachmentImpl.java:
```diff
@@ -98,12 +107,27 @@ public Optional<V> getValue(R entry) {
+
+	@Override
+	public Event<TagValueAdded<R, V>> tagValueAddedEvent() {
+		return tagValueAddedEvent;
```
```suggestion
		return this.tagValueAddedEvent;
```

LambdAurora commented on library/data/registry_entry_attachments/src/main/java/org/quiltmc/qsl/registry/attachment/impl/RegistryEntryAttachmentImpl.java:
```diff
@@ -98,12 +107,27 @@ public Optional<V> getValue(R entry) {
 	public void put(TagKey<R> entry, V value) {
+		CodecUtils.assertValid(codec, value);
 		RegistryEntryAttachmentHolder.getBuiltin(this.registry).putValue(this, entry, value);
+		tagValueAddedEvent.invoker().onTagValueAdded(entry, value);
```
```suggestion
		this.tagValueAddedEvent.invoker().onTagValueAdded(entry, value);
```

LambdAurora commented on library/data/registry_entry_attachments/src/main/java/org/quiltmc/qsl/registry/attachment/impl/RegistryEntryAttachmentImpl.java:
```diff
@@ -98,12 +107,27 @@ public Optional<V> getValue(R entry) {
+		CodecUtils.assertValid(codec, value);
+		RegistryEntryAttachmentHolder.getBuiltin(this.registry).putValue(this, entry, value,
+				BuiltinRegistryEntryAttachmentHolder.FLAG_NONE);
+		valueAddedEvent.invoker().onValueAdded(entry, value);
```
```suggestion
		this.valueAddedEvent.invoker().onValueAdded(entry, value);
```

LambdAurora commented on library/data/registry_entry_attachments/src/main/java/org/quiltmc/qsl/registry/attachment/impl/RegistryEntryAttachmentImpl.java:
```diff
@@ -98,12 +107,27 @@ public Optional<V> getValue(R entry) {
 	@Override
 	public void put(R entry, V value) {
-		RegistryEntryAttachmentHolder.getBuiltin(this.registry).putValue(this, entry, value);
+		CodecUtils.assertValid(codec, value);
```
```suggestion
		CodecUtils.assertValid(this.codec, value);
```

LambdAurora commented on library/data/registry_entry_attachments/src/main/java/org/quiltmc/qsl/registry/attachment/impl/RegistryEntryAttachmentImpl.java:
```diff
@@ -98,12 +107,27 @@ public Optional<V> getValue(R entry) {
 
 	@Override
 	public void put(TagKey<R> entry, V value) {
+		CodecUtils.assertValid(codec, value);
```
```suggestion
		CodecUtils.assertValid(this.codec, value);
```

Leo40Git committed:
this-ification

Co-authored-by: LambdAurora <aurora42lambda@gmail.com>

Leo40Git committed:
Merge branch 'rea-improvements' of github.com:/Leo40Git/quilt-standard-libraries into rea-improvements

Leo40Git committed:
Remove behavior-changing optimization

Leo40Git committed:
package-info: remove link to deleted method

Leo40Git committed:
RegistryEntryAttachment: rename getValue to get

Leo40Git commented:
(forgot to add refactor tag, since - yeah, this ~~breaks ABI~~ marks stuff for removal)

LambdAurora commented on library/data/registry_entry_attachments/src/main/java/org/quiltmc/qsl/registry/attachment/impl/RegistryEntryAttachmentImpl.java:
```diff
@@ -43,6 +46,17 @@ public RegistryEntryAttachmentImpl(Registry<R> registry, Identifier id, Class<V>
 		this.codec = codec;
 		this.side = side;
+
+		valueAddedEvent = Event.create(ValueAdded.class, listeners -> (entry, value) -> {
```
```suggestion
		this.valueAddedEvent = Event.create(ValueAdded.class, listeners -> (entry, value) -> {
```

LambdAurora commented on library/data/registry_entry_attachments/src/main/java/org/quiltmc/qsl/registry/attachment/impl/RegistryEntryAttachmentImpl.java:
```diff
@@ -43,6 +46,17 @@ public RegistryEntryAttachmentImpl(Registry<R> registry, Identifier id, Class<V>
+				listener.onValueAdded(entry, value);
+			}
+		});
+		tagValueAddedEvent = Event.create(TagValueAdded.class, listeners -> (entry, value) -> {
```
```suggestion
		this.tagValueAddedEvent = Event.create(TagValueAdded.class, listeners -> (entry, value) -> {
```

LambdAurora commented on library/data/registry_entry_attachments/src/main/java/org/quiltmc/qsl/registry/attachment/api/RegistryEntryAttachment.java:
```diff
@@ -199,7 +200,7 @@ static <R> Builder<R, String> stringBuilder(Registry<R> registry, Identifier id)
 	 * @return attachment value, or empty if no value is assigned
 	 */
-	Optional<V> getValue(R entry);
+	Optional<V> get(R entry);
```
```suggestion
	Optional<V> get(R entry);
	
	/**
	 * Gets the value associated with this attachment for the specified entry.
	 *
	 * @param entry registry entry
	 * @return attachment value, or empty if no value is assigned
	 * @deprecated use {@link #get} instead
	 */
	@Deprecated(forRemoval = true)
	default Optional<V> getValue(R entry) {
		return this.get(entry);
	}
```

Leo40Git committed:
this-ification part 2

Co-authored-by: LambdAurora <aurora42lambda@gmail.com>

Leo40Git committed:
RegistryEntryAttachment: unbreak ABI

Co-authored-by: LambdAurora <aurora42lambda@gmail.com>

Leo40Git committed:
Revert "RegistryExtensions: remove BuiltinAttachmentBuilder API"

This reverts commit 05013d04

Leo40Git committed:
RegistryExtensions: mark BuiltinAttachmentBuilder for removal

Leo40Git committed:
RegistryEntryAttachment: add getNullable

Leo40Git committed:
Validate default value

LambdAurora commented on library/data/registry_entry_attachments/src/main/java/org/quiltmc/qsl/registry/attachment/impl/BuiltinRegistryEntryAttachmentHolder.java:
```diff
@@ -0,0 +1,57 @@
+	}
+
+	public boolean isValueComputed(RegistryEntryAttachment<R, ?> attachment, R entry) {
+		return (getValueFlags(attachment, entry) & FLAG_COMPUTED) == FLAG_COMPUTED;
```
```suggestion
		return (this.getValueFlags(attachment, entry) & FLAG_COMPUTED) == FLAG_COMPUTED;
```

LambdAurora commented on library/data/registry_entry_attachments/src/main/java/org/quiltmc/qsl/registry/attachment/impl/CodecUtils.java:
```diff
@@ -0,0 +1,35 @@
+
+	public static <T> void assertValid(Codec<T> codec, T value) {
+		var result = codec.encodeStart(JsonOps.INSTANCE, value);
+		if (result.result().isEmpty()) {
```
```suggestion

		if (result.result().isEmpty()) {
```

LambdAurora commented on library/data/registry_entry_attachments/src/main/java/org/quiltmc/qsl/registry/attachment/impl/ComputedDefaultRegistryEntryAttachmentImpl.java:
```diff
@@ -39,16 +39,29 @@ public ComputedDefaultRegistryEntryAttachmentImpl(Registry<R> registry, Identifi
-			RegistryEntryAttachmentHolder.getBuiltin(this.registry).putValue(this, entry, value);
-			return Optional.of(value);
+			var encoded = codec.encodeStart(JsonOps.INSTANCE, value);
+			if (encoded.result().isEmpty()) {
```
```suggestion

			if (encoded.result().isEmpty()) {
```

LambdAurora commented on library/data/registry_entry_attachments/src/main/java/org/quiltmc/qsl/registry/attachment/impl/ConstantDefaultRegistryEntryAttachmentImpl.java:
```diff
@@ -29,11 +29,21 @@
 		super(registry, id, valueClass, codec, side);
+
+		var result = this.codec.encodeStart(JsonOps.INSTANCE, defaultValue);
+		if (result.result().isEmpty()) {
```
```suggestion

		if (result.result().isEmpty()) {
```

LambdAurora reviewed (commented):
The day I catch everything at once will be a good day.

I don't remember what happened to that ordering stuff we talked earlier though.

Leo40Git committed:
wait don't try to validate null default value

Leo40Git committed:
this-ification part 3

Co-authored-by: LambdAurora <aurora42lambda@gmail.com>

Leo40Git committed:
formatting

Leo40Git committed:
RegistryEntryAttachmentReloader: fix assets reloader never running

Forgot to remove dependency on removed Tags API reloader (RIP client tags)

Leo40Git committed:
Fix assets attachments being removed when data attachments are reloaded

Leo40Git committed:
polish testmod a bit

Leo40Git committed:
Merge remote-tracking branch 'upstream/1.19' into rea-improvements

Leo40Git committed:
Revert "RegistryEntryAttachmentReloader: fix assets reloader never running"

This reverts commit 91d6d990abba848ac0e9ee92f524a43c4aa57d9a.

Leo40Git commented:
Moved a critical bugfix to #144.

Leo40Git committed:
RegistryEntryAttachment: add iteration methods

keySet and tagKeySet for when you only need to know keys
entryIterator and tagEntryIterator for other cases

Leo40Git committed:
RegistryEntryAttachment.put: add notice about when to call

Leo40Git committed:
fix weird double newline that Oro accidentally did

(according to git blame, anyway)

Leo40Git committed:
make entryIterator and tagEntryIterator work as expected

Leo40Git committed:
Add test for iteration methods

Leo40Git committed:
DispatchedAttachmentTest: make tag messages consistent with entry messages

EnnuiL commented on library/data/registry_entry_attachments/src/main/java/org/quiltmc/qsl/registry/attachment/api/RegistryEntryAttachment.java:
```diff
@@ -261,6 +323,102 @@ public boolean shouldLoad(ResourceType source) {
+	 * @param <R> type of registry entry
+	 * @param <V> type of value
+	 */
+	@SuppressWarnings("ClassCanBeRecord")
```
I'm curious about why these (and other similar classes) aren't records

Leo40Git replied:
Honestly I don't remember why I have a vendetta against records either

Leo40Git replied:
nvm, I remembered: you can't document the getters
It's probably fine, though

EnnuiL replied:
That makes kinda sense? Although yeah, I don't really see a point on documenting with the very boilerplate-y "returns the X"; I feel like the comments on the record itself already helps, although you did manage to pull it off anyway, so perhaps this isn't much of an issue anyway

EnnuiL commented on library/data/registry_entry_attachments/src/testmod/java/org/quiltmc/qsl/registry/attachment/test/DispatchedAttachmentTest.java:
```diff
@@ -91,6 +101,28 @@ public TypedActionResult<ItemStack> use(World world, PlayerEntity user, Hand han
+
+	@Override
+	public void onEndDataPackReload(@Nullable MinecraftServer server, ResourceManager resourceManager, @Nullable Throwable error) {
+		if (error != null) {
```
Nitpick: perhaps inline the return?

EnnuiL commented on library/data/registry_entry_attachments/src/main/java/org/quiltmc/qsl/registry/attachment/impl/BuiltinRegistryEntryAttachmentHolder.java:
```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright 2021-2022 QuiltMC
```
Nitpick:
```suggestion
 * Copyright 2022 QuiltMC
```

EnnuiL reviewed (commented):
I just remembered that I've written down a really basic draft of a review and then abandoned it; I don't have any other complaints *and* perhaps this review should be fine

Leo40Git committed:
Correct copyright years on new files

Leo40Git committed:
DispatchedAttachmentTest: inline return

Leo40Git committed:
RegistryEntryAttachment: convert Entry and TagEntry to records

man, javadoc on records is *weird*
also IDEA fucked the indents when it converted the classes

Leo40Git committed:
RegistryEntryAttachmentSync: oro pls remember to clear tags

Leo40Git committed:
oops fix build

Leo40Git committed:
changed the wrong class' license

Leo40Git committed:
Merge branch '1.19' of https://github.com/QuiltMC/quilt-standard-libraries into rea-improvements

Leo40Git committed:
RegistryEntryAttachmentImpl: optimize EntryIterator.next()
